Map matcher funcs

nelsam opened the pull request:
My initial proposal for adding MatcherFuncs to MapController, MapStatic, and MapStaticFile.

sam-nelson-mcn committed:
Added matcherFuncs to mapper methods that didn't support them

Step 2 (because I like doing TDD backwards, I guess): Make the required changes.  Now I get to go to step 1 and create the tests.

sam-nelson-mcn committed:
Testing Controller with MatcherFuncs

Step 1: Testing.
I added tests for mapping a Controller with MatcherFuncs.  I'm guessing it needs some work (I am at the end of my work day and don't have time to look in to testing messages, but I'm guessing I need to change some strings in the tests), but it's something.  I'll toy around with it a little later and see what happens.

nelsam committed:
Fixed MapController HandlerFunc tests

I had some bad syntax in my first testing commit.  It should be fixed now.

nelsam committed:
Fixed my varargs calls

I didn't realize until just this moment that ...args in go isn't quite as easy as *args in python.  I can do func(args...) or func(arg1, arg2, arg3), but not func(arg1, args...).

So I got rid of that at the end of my calls.  Maybe I'll just update handlerForOptions to handle an array.

nelsam committed:
Updated handlerForOptions to accept arguments of type []MatcherFunc

Because of the extra code it would take (and how ugly that extra code would be) to create a varargs slice of type []interface{} that contained all of the values that need to be passed to Map, MapBefore, and MapAfter from within MapController, I thought it might be a better idea to modify handlerForOptions to accept arguments of type []MatcherFunc as well as MatcherFunc.  This allows me to just generate the matcherFuncs array once in MapController, then pass it on as a normal argument to Map.

nelsam committed:
loadMatcherFuncs function

I had nearly identical logic in two places, so I refactored it into a function that could be called in both places.

matryer commented on handlers/mapping.go:
```diff
@@ -21,6 +21,25 @@ var (
 
+// Look for supported MatcherFunc option types and return a full list of
+// all MatcherFuncs found.
+func loadMatcherFuncs(options ...interface{}) []MatcherFunc {
```
Consider rename to `findMatcherFuncs` since it doesn't load anything.

matryer commented on handlers/mapping_test.go:
```diff
@@ -341,6 +341,29 @@ func TestMapController(t *testing.T) {
+	codecService := new(codecservices.WebCodecService)
+	handler := NewHttpHandler(codecService)
+
+	matcherFunc := MatcherFunc(func(ctx context.Context) (MatcherFuncDecision, error) {
```
I think this should also work without casting to `MatcherFunc`.  So, have `func(context.Context) (MatcherFuncDecision, error)` as an option in the case switch.

sam-nelson-mcn committed:
Renamed loadMatcherFuncs to findMatcherFuncs and support func(context.Context) (MatcherFuncDecision, error)

I renamed loadMatcherFuncs to findMatcherFuncs, since it doesn't actually do any loading of anything.

findMatcherFuncs now supports arguments of type func(context.Context) (MatcherFuncDecision, error), as well, so that users don't have to cast their matcher functions to MatcherFunc before passing them in.

I created a test that explicitly tests the three types of arguments that findMatcherFuncs should accept, to make absolutely sure that it is working.
